src/Datadog.Trace.ClrProfiler.Managed/CallTarget/Handlers/BeginMethodHandler.cs

```diff
@@ -0,0 +1,41 @@
+using System;
+using System.Reflection.Emit;
+using System.Runtime.CompilerServices;
+#pragma warning disable SA1649 // File name must match first type name
+
+namespace Datadog.Trace.ClrProfiler.CallTarget.Handlers
+{
+    internal static class BeginMethodHandler<TIntegration, TTarget>
+    {
+        private static readonly InvokeDelegate _invokeDelegate;
+
+        static BeginMethodHandler()
+        {
+            try
+            {
+                DynamicMethod dynMethod = IntegrationMapper.CreateBeginMethodDelegate(typeof(TIntegration), typeof(TTarget), Util.ArrayHelper.Empty<Type>());
```

Comment:
Is my understanding correct that we will require one type for each integration method we want to call? For instance, DbCommand.ExecuteReader and DbCommand.ExecuteReaderAsync would each need a different type? Or could they share the same `OnMethodBegin` methods? I apologize if this is shown in the implementation later (feel free to point this out to me), but so far I am at the end of reviewing all of the managed C# code and will soon move on to the C++ code

Reply:
Each integration will create a custom `DynamicMethod` to convert the `CallTargetInvoker` call to the actual method signature, this is done to handle the possible conversions from a `TArg` to an actual type defined in the integration, and to enable the ducktyping mechanism based on the constraints defined by the integration. Each integration could possible define different ways to be called.

The implementation doesn't share those DynamicMethods and uses `TIntegration` as a generic parameter for performance reasons. By storing this in a static generic class we remove a look up operation to a collection in order to get the delegate, so the runtime ensure that if we use this generic class with the same `TIntegration`, `TTarget`, and `TArgs`, we will have the exact delegate to make a successful call to the integration definition.

That's the whole reason to use statics+generics extensively in the implementation, we have a cheap cache system built in.